### ISSUE TITLE
LaTeX Error: Option clash for package caption (when making with pdflatex with texlive-2008)

### ISSUE THREAD

**Issue by imz** [rominf/latex-g7-32#23]
Когда транслируется пример с помощью pdflatex с texlive-2008 из ALT Sisyphus (да, старый; может быть с новыми версиями такой ошибки и не будет) возникает ошибка:

    ! LaTeX Error: Option clash for package caption.
   
    See the LaTeX manual or LaTeX Companion for explanation.
    Type  H <return>  for immediate help.
     ...                                              
    
    l.7 \include{preamble.inc}
    
    \@input{listings.inc.aux}

Подробности я записал в <https://bugzilla.altlinux.org/show_bug.cgi?id=31414#c1>.

Но в результате PDF какой-никакой получается.

**Comment by rominf** [comment]
Да, пакет `caption` очень проблемный. Вообще говоря, по хорошему от него надо избавиться, но из-за чего-то у меня не получилось.

**Comment by imz** [comment]
Столкновение разного набора опций пакета при его загрузке происходит между двумя местами: в `preamble.inc.tex` и `G2-105.sty`. Первое место недавно (2015-04-12) вставил Роман, второе -- раньше (2010-06-06) Иван Коротков. Так что автору (Роману) должно быть лучше понятно, какого эффекта нужно было добиться и можно ли заменить на какой-нибудь `\captionsetup` (и куда смотреть в примере, чтобы этот эффект увидеть и проверить). Был бы благодарен за пояснения Роману ради приведения примера к чисто собирающемуся виду.

Clash между <https://github.com/rominf/latex-g7-32/blob/master/tex/preamble.inc.tex#L49>

   % Центрирование подписей к плавающим окружениям
    \usepackage[justification=centering]{caption}

и <https://github.com/rominf/latex-g7-32/blob/master/tex/G2-105.sty#L399>:

    % Подпись по умолчанию слева (для таблиц, листингов и т.п.)
    \RequirePackage[normal,nooneline]{caption}
    % Длинное тире в качестве разделителя
    \DeclareCaptionLabelSeparator*{emdash}{\space\textemdash\space}
    % Между подписью и рисунком 10pt
    \captionsetup{labelsep=emdash,aboveskip=10pt,belowskip=0pt,position=bottom}
    % У рисунков вырвнивание по центру
    \captionsetup[figure]{justification=centering}
    % У таблиц -- слева, зазор 5pt вместо 10
    \captionsetup[table]{position=top,aboveskip=5pt,justification=justified}

Про такие clash ещё пишут:

* <http://tex.stackexchange.com/q/77895/3463>
* <http://compgroups.net/comp.text.tex/option-clash-with-package-caption/1925853>
* <http://www.macfreek.nl/memory/LaTeX_package_conflicts#Caption_and_Subfig>

**Comment by rominf** [comment]
ОК, спасибо. С радостью приму PR, решающий проблему, либо решу через неопределённое кол-во времени (см. #21).

**Comment by imz** [comment]
Ага, про древний texlive в ALTLinux. (Я попозже может впишу сюда ещё ссылки на обсуждение этого.) Но ситуация такая, что превращать в rpm тысячи texlive-овских пакетов -- дело для человека плохое, лучше (если уж надо) делать роботом (были попытки). К тому же есть такое общее соображение/проблема:

чем тащить своими силами кучу пакетов, которая живёт своей жизнью, в дистрибутив и запаздывать и не решать все потребности, лучше пользователю посоветовать пользоваться родным специфичным менеджером пакетов. (Скажем, tlmgr для texlive, cabal/что-то ещё для Haskell чем тащить "cabal hell"  с hackage в дистрибутив -- разные библиотеки требуют разных версий друг друга; родные средства для Go. Я просто об этом задумывался в последние дни, поэтому охотно пишу.)

Для чего же тогда пакеты в дистрибутиве?

Разные люди отвечают: чтобы собирать другие пакеты.

Про TeX такое говорилось в обсуждениях в ALTLinux когда-то (главное, чтобы всю документацию собирал; пока собирает, можно оставить старое патченное).

Про пакеты golang прочитал такое в политиках упаковки пакетов Go для Fedora.

Про Haskell я сам к каким-то таким выводам пришёл, совершая попытку понять, что делать со старым ghc в ALTLinux.

В ALTLinux есть много работы по созданию роботов для вливания таких репозиториев в дистрибутив. (Попытка про texlive: <http://autoextra.altlinux.org/pub/ALTLinux/texlive/Sisyphus/files/list/bin.list>)

Ещё мне кажется привлекательным подход NixOS, как у них получается быстро превращать сторонние языкоспецифичные репозитории под контроль единого пакетного менеджера nix (единый -- лучше, чем много разных для каждого языка, потому что можно хорошо отслеживать взаимозависимости между пакетами разного происхождения).

**Comment by imz** [comment]
Я бы пропатчил для решения проблемы, мне просто помог бы совет, чего добавленная Вами строчка в preamble должна добиваться. На что смотреть в примере, чтобы проверить.

figure и table настраиваются в G2-105.sty

А вот Ваша:

    % Центрирование подписей к плавающим окружениям
    \usepackage[justification=centering]{caption}

задумана как влияющая на те настройки, или для других объектов (всех остальных? А какие ещё есть в примере с подписями?)

Если быстрого ответа у Вас сейчас нет, не страшно, я обдумаю эти дела спустя какое-то время.

**Comment by rominf** [comment]
Специальные пакетные менеджеры - это хорошо, проблема в разных API (пользователю надо помнить `zypper rm` vs `pip uninstall`) и том, что каждый раз пишут велосипед.

Насчёт конкретно tex - в openSUSE написали какое-то средство для этого очевидно. Сколько сил и времени они потратили не знаю, но пользоваться очень удобно.

Насчёт latex-g7-32: проект писался большим количеством людей в разное время, я сам до конца не понимаю его логику в некоторых местах, по поводу этих строк я не помню, надо вникать и экспериментировать.
